core/domain/user_services.py

```diff
@@ -680,36 +680,72 @@ def _save_user_settings(user_settings):
-        last_started_state_editor_tutorial=(
-            user_settings.last_started_state_editor_tutorial),
-        last_started_state_translation_tutorial=(
-            user_settings.last_started_state_translation_tutorial),
-        last_logged_in=user_settings.last_logged_in,
-        last_edited_an_exploration=user_settings.last_edited_an_exploration,
-        last_created_an_exploration=(
-            user_settings.last_created_an_exploration),
-        profile_picture_data_url=user_settings.profile_picture_data_url,
-        default_dashboard=user_settings.default_dashboard,
-        creator_dashboard_display_pref=(
-            user_settings.creator_dashboard_display_pref),
-        user_bio=user_settings.user_bio,
-        subject_interests=user_settings.subject_interests,
-        first_contribution_msec=user_settings.first_contribution_msec,
-        preferred_language_codes=user_settings.preferred_language_codes,
-        preferred_site_language_code=(
-            user_settings.preferred_site_language_code),
-        preferred_audio_language_code=(
-            user_settings.preferred_audio_language_code),
-        deleted=user_settings.deleted
-    ).put()
+    user_model = user_models.UserSettingsModel.get_by_id(user_settings.user_id)
+
+    # If user with the given user_id already exists, update that model
+    # with the given user settings, otherwise, create a new one.
+    if user_model is not None:
+        user_model.populate(
+            email=user_settings.email,
+            role=user_settings.role,
+            username=user_settings.username,
+            normalized_username=user_settings.normalized_username,
+            last_agreed_to_terms=user_settings.last_agreed_to_terms,
+            last_started_state_editor_tutorial=(
+                user_settings.last_started_state_editor_tutorial),
+            last_started_state_translation_tutorial=(
+                user_settings.last_started_state_translation_tutorial),
+            last_logged_in=user_settings.last_logged_in,
+            last_edited_an_exploration=user_settings.last_edited_an_exploration,
+            last_created_an_exploration=(
+                user_settings.last_created_an_exploration),
+            profile_picture_data_url=user_settings.profile_picture_data_url,
+            default_dashboard=user_settings.default_dashboard,
+            creator_dashboard_display_pref=(
+                user_settings.creator_dashboard_display_pref),
+            user_bio=user_settings.user_bio,
+            subject_interests=user_settings.subject_interests,
+            first_contribution_msec=user_settings.first_contribution_msec,
+            preferred_language_codes=user_settings.preferred_language_codes,
+            preferred_site_language_code=(
+                user_settings.preferred_site_language_code),
+            preferred_audio_language_code=(
+                user_settings.preferred_audio_language_code),
+            deleted=user_settings.deleted
```

Comment:
Could you put all these values into a dictionary and then just unpack it here by using `user_model.populate(**your_dict)`. In the `else` branch you can add `id` and `gae_id` to the dictionary and unpack it there like `user_models.UserSettingsModel(**your_dict)`.

Reply:
Done, thanks!